tests/ctrl/services.go

```diff
@@ -0,0 +1,132 @@
+package ctrl
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/Inphi/eip4844-interop/shared"
+	"github.com/ethereum/go-ethereum/ethclient"
+	beaconservice "github.com/prysmaticlabs/prysm/v3/proto/eth/service"
+	"google.golang.org/grpc"
+)
+
+type Service interface {
+	Start(ctx context.Context) error
+	Stop(ctx context.Context) error
+	Started() <-chan struct{}
+}
+
+func NewBeaconNode() Service {
+	url := fmt.Sprintf("http://%s/eth/v1/beacon/genesis", shared.BeaconGatewayGRPC)
+	return newDockerService("beacon-node", url)
+}
+
+func NewValidatorNode() Service {
+	return newDockerService("validator-node", shared.ValidatorRPC)
+}
+
+func GetBeaconNodeClient(ctx context.Context) (beaconservice.BeaconChainClient, error) {
+	// TODO: cache conns for reuse
+	conn, err := grpc.DialContext(ctx, shared.BeaconRPC, grpc.WithInsecure())
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to dial beacon grpc", err)
+	}
+	return beaconservice.NewBeaconChainClient(conn), nil
+}
+
+func NewBeaconNodeFollower() Service {
+	url := fmt.Sprintf("http://%s/eth/v1/beacon/genesis", shared.BeaconGatewayFollowerGRPC)
+	return newDockerService("beacon-node-follower", url)
+}
+
+func GetBeaconNodeFollowerClient(ctx context.Context) (beaconservice.BeaconChainClient, error) {
+	// TODO: cache conns for reuse
+	conn, err := grpc.DialContext(ctx, shared.BeaconFollowerRPC, grpc.WithInsecure())
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to dial beacon follower grpc", err)
+	}
+	return beaconservice.NewBeaconChainClient(conn), nil
+}
+
+func NewGethNode() Service {
+	return newDockerService("execution-node", shared.GethRPC)
+}
+
+func NewGethNode2() Service {
+	return newDockerService("execution-node-2", shared.GethRPC)
+}
+
+func GetExecutionClient(ctx context.Context) (*ethclient.Client, error) {
+	client, err := ethclient.DialContext(ctx, shared.GethRPC)
+	if err != nil {
+		return nil, fmt.Errorf("%w: Failed to connect to the Ethereum client", err)
+	}
+	return client, nil
+}
+
+type dockerService struct {
+	started   chan struct{}
+	svcname   string
+	statusURL string
+}
+
+func (s *dockerService) Start(ctx context.Context) error {
+	if err := StartServices(s.svcname); err != nil {
+		return err
+	}
+	req, err := http.NewRequestWithContext(ctx, "GET", s.statusURL, nil)
```

Comment:
Is there some risk here that the service needs more time to start before it can accept status requests and will return error even though it might be fine, just not quite ready yet?

Reply:
`http.NewRequestWithContext` doesn't do any I/O. It only fails on invalid inputs (ex: providing an unknown HTTP method).

Reply:
Yes, but right after this (L83) you then execute the request (L83) which does do I/O?  Oh I see, it loops until there is no error. Can you put a comment there explaining what's going on?